nmigen_soc/wishbone/bus.py

```diff
@@ -166,6 +167,147 @@ def memory_map(self, memory_map):
+                            .format(sub_bus))
+        intr_size = (2**intr_bus.addr_width)*intr_bus.data_width
+        sub_size = (2**sub_bus.addr_width)*sub_bus.data_width
+        if intr_size != sub_size:
+            raise ValueError("Total bit size of initiator and subordinate bus have to be the same")
+        if sub_bus.granularity > intr_bus.granularity:
+            raise ValueError(
+                "Granularity of subordinate bus has to be smaller or equal to "
+                "granulariy of initiator bus"
+            )
+        for opt_output in {"lock"}:
+            if hasattr(intr_bus, opt_output) and not hasattr(sub_bus, opt_output):
+                raise ValueError("Initiator bus has optional output {!r}, but the suborbdinate bus "
+                                 "does not have a corresponding input"
+                                 .format(opt_output))
+        for opt_output in {"err", "rty"}:
+            if hasattr(sub_bus, opt_output) and not hasattr(intr_bus, opt_output):
+                raise ValueError("Subordinate bus has optional output {!r}, but the initiator bus "
+                                 "does not have a corresponding input"
+                                 .format(opt_output))
+        if intr_bus.data_width > sub_bus.data_width:
+            raise NotImplementedError(
+                "Support for multi-cycle bus operation when initiator data_width is"
+                "bigger than the subordinate one is not implemented."
+            )
+
+        self.intr_bus = intr_bus
+        self.sub_bus = sub_bus
+        self.domain = domain
+
+    def elaborate(self, platform):
+        intr_bus = self.intr_bus
+        sub_bus = self.sub_bus
+
+        m = Module()
+
+        common_addr_width = min(intr_bus.addr_width, sub_bus.addr_width)
+        m.d.comb += [
+            sub_bus.cyc.eq(intr_bus.cyc),
+            sub_bus.we.eq(intr_bus.we),
+            sub_bus.adr[(sub_bus.addr_width-common_addr_width):sub_bus.addr_width].eq(
+                intr_bus.adr[(intr_bus.addr_width-common_addr_width):intr_bus.addr_width]
+            ),
+            intr_bus.ack.eq(sub_bus.ack),
+        ]
+        if hasattr(intr_bus, "err"):
+            m.d.comb += intr_bus.err.eq(getattr(sub_bus, "err", 0))
+        if hasattr(intr_bus, "rty"):
+            m.d.comb += intr_bus.rty.eq(getattr(sub_bus, "rty", 0))
+        if hasattr(sub_bus, "lock"):
+            m.d.comb += sub_bus.lock.eq(getattr(intr_bus, "lock", 0))
+        if hasattr(sub_bus, "cti"):
+            m.d.comb += sub_bus.cti.eq(getattr(intr_bus, "cti", CycleType.CLASSIC))
+        if hasattr(sub_bus, "bte"):
+            m.d.comb += sub_bus.bte.eq(getattr(intr_bus, "bte", BurstTypeExt.LINEAR))
+
+
+        # Following code assumes single cycle read and write;
+        # duplicate check that was also done in __init__()
+        assert intr_bus.data_width <= sub_bus.data_width, "Internal error: should not happed"
```

Comment:
Redundancy aside, it would be preferable to check properties early in `__init__()`, if possible. Also, `intr_bus` and `sub_bus` are user provided, so exceptions make more sense here.

Reply:
Removed it. Check is already done in __init__() and in hindsight extra check is not needed.